tasks/solve-a-hidato-puzzle/src/main.rs

```diff
@@ -0,0 +1,213 @@
+use std::cmp::{max, min};
+use std::fmt;
+use std::ops;
+
+#[derive(Debug, Clone, PartialEq)]
+struct Board {
+    cells: Vec<Vec<Option<u32>>>,
+}
+
+impl Board {
+    fn new(initial_board: Vec<Vec<u32>>) -> Self {
+        let b = initial_board
+            .iter()
+            .map(|r| {
+                r.iter()
+                    .map(|c| if *c == u32::MAX { None } else { Some(*c) })
+                    .collect()
+            })
+            .collect();
+
+        Board { cells: b }
+    }
+
+    fn height(&self) -> usize {
+        self.cells.len()
+    }
+
+    fn width(&self) -> usize {
+        self.cells[0].len()
+    }
+}
+impl ops::Index<(usize, usize)> for Board {
+    type Output = Option<u32>;
+
+    fn index(&self, (y, x): (usize, usize)) -> &Self::Output {
+        &self.cells[y][x]
+    }
+}
+impl ops::IndexMut<(usize, usize)> for Board {
+    /// Returns a mutable reference to an cell for a given 'x' 'y' coordinates
+    fn index_mut(&mut self, (y, x): (usize, usize)) -> &mut Option<u32> {
+        &mut self.cells[y][x]
+    }
+}
+
+impl fmt::Display for Board {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let output: Vec<String> = self
+            .cells
+            .iter()
+            .map(|r| {
+                let mut row = String::default();
+
+                r.iter().for_each(|c| match c {
+                    None => row.push_str(format!("{:>2} ", " ").as_ref()),
+                    Some(c) if c == &0 => row.push_str(format!("{:>2} ", ".").as_ref()),
+                    Some(c) => row.push_str(format!("{:>2} ", c).as_ref()),
+                });
+                row
+            })
+            .collect();
+
+        write!(f, "{}", output.join("\n"))
+    }
+}
+
+/// Structure for holding puzzle related information
+/// * `board`: the state of the board
+/// * `fixed`: all the numbers which were given at puzzle setup,
+/// aka all the numbers which cannot be changed during solving the puzzle
+/// * `start`: position of the first number (1)
+#[derive(Clone, Debug)]
+struct Puzzle {
+    board: Board,
+    fixed: Vec<u32>,
+    start: (usize, usize),
+}
```

Comment:
```suggestion
/// Structure for holding puzzle related information.
#[derive(Clone, Debug)]
struct Puzzle {
    /// The state of the board.
    board: Board,
    
    /// All the numbers which were given at puzzle setup:
    /// the numbers which cannot be changed during solving the puzzle.
    fixed: Vec<u32>,
    
    /// Position of the first number (1).
    start: (usize, usize),
}
```

The field comments should be written on the fields themselves.